tls/s2n_handshake_io.c

```diff
@@ -378,13 +380,27 @@ static message_type_t tls13_handshakes[S2N_HANDSHAKES_COUNT][S2N_MAX_HANDSHAKE_L
             APPLICATION_DATA
     },
 
+    [NEGOTIATED | WITH_EARLY_DATA] = {
+            CLIENT_HELLO,
+            SERVER_HELLO, ENCRYPTED_EXTENSIONS, SERVER_FINISHED,
+            END_OF_EARLY_DATA, CLIENT_FINISHED,
+            APPLICATION_DATA
+    },
+
     [NEGOTIATED | MIDDLEBOX_COMPAT] = {
             CLIENT_HELLO,
             SERVER_HELLO, SERVER_CHANGE_CIPHER_SPEC, ENCRYPTED_EXTENSIONS, SERVER_FINISHED,
             CLIENT_CHANGE_CIPHER_SPEC, CLIENT_FINISHED,
             APPLICATION_DATA
     },
 
+    [NEGOTIATED | MIDDLEBOX_COMPAT | WITH_EARLY_DATA] = {
+            CLIENT_HELLO, CLIENT_CHANGE_CIPHER_SPEC,
```

Comment:
Why did you move CLIENT_CHANGE_CIPHER_SPEC to be after the CLIENT_HELLO?

Reply:
I reference the RFC about this in the tests. I didn't want to start adding compliance comments to handshakes, since in most cases you'd need a bunch of comments on each handshake :) 


> If not offering early data, the client sends a dummy
> change_cipher_spec record (see the third paragraph of Section 5)
> immediately before its second flight.  This may either be before
> its second ClientHello or before its encrypted handshake flight.
> **If offering early data, the record is placed immediately after the
> first ClientHello.**